orttraining/orttraining/python/training/ortmodule/ort_triton/_utils.py

```diff
@@ -0,0 +1,172 @@
+# -------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License.
+# --------------------------------------------------------------------------
+
+import re
+import uuid
+from collections import defaultdict
+from typing import Any, List, Tuple
+
+import numpy as np
+from onnx import GraphProto, NodeProto, TensorProto, helper, numpy_helper
+
+
+def gen_unique_name(prefix: str) -> str:
+    return prefix + "_" + uuid.uuid4().hex[:8]
+
+
+def _topological_soft_internal(node, visited, output_consumers, sorted_nodes):
+    visited.add(node.name)
+    for next_node in output_consumers[node.name]:
+        if next_node.name not in visited:
+            _topological_soft_internal(next_node, visited, output_consumers, sorted_nodes)
+
+    sorted_nodes.insert(0, node)
+
+
+# Topological sort of nodes given the input names. The list of nodes contain both constant and non-constant nodes.
+def topological_sort(inputs: List[str], nodes: List[NodeProto]) -> List[NodeProto]:
+    const_nodes = []
+    non_const_nodes = []
+    for node in nodes:
+        if not node.name:
+            node.name = gen_unique_name(node.op_type)
+        if node.op_type == "Constant":
+            inputs.append(node.output[0])
+            const_nodes.append(node)
+        else:
+            non_const_nodes.append(node)
+
+    # Build relationship between nodes.
+    graph_input_consumers = defaultdict(list)
+    output_consumers = defaultdict(list)
+    input_set = set(inputs)
+    for node in non_const_nodes:
+        for input in node.input:
+            if input in input_set:
+                graph_input_consumers[input].append(node)
+        for output in node.output:
+            if not output:
+                continue
+            for consumer in non_const_nodes:
+                if output in consumer.input:
+                    output_consumers[node.name].append(consumer)
+
+    # Topological sort.
+    visited = set()
+    sorted_nodes = []
+    for input in inputs:
+        for node in graph_input_consumers[input]:
+            if node.name not in visited:
+                _topological_soft_internal(node, visited, output_consumers, sorted_nodes)
+
+    return const_nodes + sorted_nodes
+
+
+# Get attribute value from node by attribute key.
+def get_attribute(node: NodeProto, attr_name: str, default_value: Any = None) -> Any:
+    found = [attr for attr in node.attribute if attr.name == attr_name]
+    if found:
+        return helper.get_attribute_value(found[0])
+    return default_value
+
+
+# Convert Constant node or TensorProto to numpy array.
+def to_numpy_array(node: Any) -> np.ndarray:
+    tensor = node
+    if isinstance(node, NodeProto):
+        tensor = get_attribute(node, "value")
+    assert isinstance(tensor, TensorProto)
+    return numpy_helper.to_array(tensor)
+
+
+_TENSOR_TYPE_TO_NP_TYPE = {
```

Comment:
This is already defined in onnx package.

Reply:
I've changed the code to use the definition from ONNX. But I found that the latest main of ONNX changed TENSOR_TYPE_TO_NP_TYPE to TENSOR_TYPE_MAP with different data structure. Reusing definition from some other packages seems like we can save duplicated code, but in reality it brings maintain trouble, especially when different users use different versions of packages.